feat: lerna project

andrestone opened the pull request:
*Issue #, if available:*
- closes https://github.com/eladb/projen/issues/9

see README.

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

eladb commented on src/lerna.ts:
```diff
@@ -0,0 +1,93 @@
+  // new project with outdir on the package location, validate workspaces and locations using minimatch
+  public addPackage(name: string, location: string, project: NodeProject) {
+    // Run validation for this._options.rootScripts
+    const packageJson = Node.of(project).tryFindChild('package.json') as JsonFile; // This or expose props on NodeProject ?
```
Expose props

andrestone commented:
I will. It's that I needed to poke around a bit first in order to get a better feel.

eladb commented on src/lerna.ts:
```diff
@@ -0,0 +1,93 @@
+
+    // We need to somehow change the outdir...
+    try {
+      project.synth({outdir: location});
```
`synth` should be done as part of the parent project's `synth()`. Otherwise, mutations to the sub-project will be lost.

andrestone replied:
Hmm... What if it isn't part of the root construct? I'm thinking of it like a `nestedStack`, which follows a different `synth` route. It makes it easier to manipulate the `outdir`. 

Another option would be having it as part of the tree but then I'm not sure about what the logic would be to make it synth to a different outdir (maybe extending a `FileBase` just for Lerna or go passing props up?).

As a third option would be having a `.projenrc.js` file created on the `location` and having projen to calling itself there. But this one feels pretty bad.

eladb commented on src/lerna.ts:
```diff
@@ -0,0 +1,93 @@
+    this._packages.push(project);
+  }
+
+  public addBootstrappedPackage(name: string, location: string, npxCommand: string, createDir?: boolean) {
```
What's that? doc strings will help with the review

andrestone replied:
README incoming

andrestone commented:
But if you could comment at least on the two points of the TODO, it would really help me organize my thoughts.

eladb commented on src/lerna.ts:
```diff
@@ -0,0 +1,93 @@
+    this.addPackage(name, location, NodeProject.fromExternalScript(npxCommand, name, createDir))
+  }
+
+  public async addPackageFromTemplate(name: string, location: string, templateName: string) {
```
?? Not sure I understand the use case / workflow here.

eladb commented on src/lerna.ts:
```diff
@@ -0,0 +1,93 @@
+  /**
+   * Packages to create
+   */
+  readonly packages: [{name: string, location: string, project: NodeProject}];
```
Feels like `location` should be optional and default to `packages/$name`, no?

eladb commented on src/lerna.ts:
```diff
@@ -0,0 +1,93 @@
+   * Projen will fail if any script is missing on any of the packages.
+   *
+   */
+  readonly rootScripts?: string[];
```
This should be have a similar API and funnel into the same path that manages scripts in `NodeProject`.

andrestone replied:
The intention here is just to ensure the scripts exist in the packages. Is it worth it to have a method to add identical scripts to all child packages?

eladb commented on src/node-project.ts:
```diff
@@ -367,7 +373,7 @@ export class NodeProject extends Project {
       });
 
-      if (options.releaseToNpm) {
+      if (options.releaseToNpm && !options.private) {
```
I thought we already had support for `private`. At any rate, shouldn't that also go into `package.json`?

eladb commented on src/node-project.ts:
```diff
@@ -544,6 +550,14 @@ export class NodeProject extends Project {
     );
   }
+
+  public static fromExternalScript(npxCommand: string, name: string, createDir?: boolean): NodeProject {
```
I am failing to understand what this is about

andrestone replied:
Moved to NodeProject (README).

eladb commented:
> Create outdir parameter for Project.synth() / another solution for synthesizing to a different outdir

I think something like that would work, yes. Also see my comment about centralized synthesis.

> Interface with NodePorject use construct node to get the JsonFile object or expose NodeProject properties?

Yes, add properties to expose the various things. `package.json`, by the way, can already be controlled via `project.manifest` (bad name, yes).

> Logic for projects from scripts (should this belong to a separate PR? Could be generally used to bootstrap a simple NodeProject using the provided external script and projen overrides)

Yes, sounds like a separate PR, and I am not sure I fully understand the workflow around it, so worth a discussion before you dive into it. At any rate, I would not include that here.

andrestone commented:
> Yes, add properties to expose the various things. `package.json`, by the way, can already be controlled via `project.manifest` (bad name, yes).

Should I always try exposing a property as a first solution?

andrestone commented:
I have included a first README. I know this should be splitted in two PR's, please share your thoughts on if and where it's worth doing so.

eladb commented on README.md:
````diff
@@ -177,10 +177,221 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+```typescript
+const { NodeProject } = require('projen');
+
+const cdkConstruct = NodeProject.fromExternalScript({
````
I don't like this api and I don't see how this is related to lerna. Can we please get rid of it for now?

andrestone replied:
Yep! It isn't really related, I just hijacked the PR to get your thoughts on it (sorry).

eladb commented on README.md:
```diff
@@ -177,10 +177,221 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+  /* ... */
+});
+
+const myService = NodeProject.fromTemplate({
```
Why is this needed?

eladb replied:
Templates are only a way to bootstrap a projenrc file

andrestone replied:
This isn't needed, it's a related feature. I'll just discard it and add to another PR, as suggested.

eladb commented on README.md:
```diff
@@ -177,10 +177,221 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+      "packages/@iac/*",
+      "packages/@services/*",
+    ],
+    nohoist: [],
```
Example?

I actually think this api is too low level. Can we provide some higher level support like `lerna.noHoist(package, dep)`

eladb commented on README.md:
```diff
@@ -177,10 +177,221 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+const lernaProject = new LernaProject({
+  name: 'my-monorepo-project',
+  workspaces: {
+    packages: [
```
This feels redundant since we know exactly where are all the packages so we should generate this based
On the packages added to the project and not through a wildcard.

andrestone replied:
The idea here is to just have a way to validate package locations vs workspaces paths. How could we validate it without declared workspace patterns? Should we just add any arbitrary package location as a workspace?

eladb commented on README.md:
```diff
@@ -177,10 +177,221 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+    nohoist: [],
+  packages: [
+    {
+      location: 'packages/@services',
```
This should be optional and default to `packages/PROJECTNAME`

andrestone replied:
It is described like so.

eladb commented on README.md:
```diff
@@ -177,10 +177,221 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+  /* ... */
+});
+
+lernaProject.addPackage({
```
The minimum api should be 

```
lerna.addPackage(project);
```

eladb commented on README.md:
```diff
@@ -177,10 +177,221 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+>will default to `packages`. If you're willing to use the default location when adding packages and 
+>want to specify additional packages location patterns, don't forget to include `packages/*`.
+
+#### Using Projen templates and external scripts
```
Please get rid of this feature for now. Let's discuss later

eladb commented on README.md:
```diff
@@ -177,10 +177,221 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+
+More details on templates and external scripts [here](#templates-and-external-scripts). 
+
+#### Enforcing the use of root level scripts
```
Let's remove this feature for now and add later. Ok?

In the meantime users can easily achieve this by passing in the same set of scripts to all projects (through a local variable)

andrestone commented:
- Requested
- Version file is now created during synth
- Preserve field functionality for `JsonFile` to support the above (array of fields not to overwrite on `synth`)
- `LernaVersion` to *optionally* get version for package.json via `lerna.json` and not create `.versionrc.json`
- Moved `main` key definition out of `NodeProject` (to jsii / typescript)
- Cleaned up package.json output a bit
- First iteration of a cdk8s monorepo mock `examples/lerna`

eladb commented on README.md:
```diff
@@ -177,11 +177,93 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+  /* ... */
+});
+
+lernaProject.addPackage(new TypeScriptLibraryProject({ 
+    name: 'geocoding-service', 
+    description: 'Geocoding Service' 
+  }),
+);
```
Style:

```suggestion
lernaProject.addPackage(new TypeScriptLibraryProject({ 
  name: 'geocoding-service', 
  description: 'Geocoding Service' 
}));
```

eladb commented on README.md:
```diff
@@ -177,11 +177,93 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+});
+lernaProject.addPackage(myChildProject);
+lernaProject.noHoist('any-package', 'yaml') // generates 'any-package/yaml' nohoist entry
+lernaProject.noHoist(myChildProject, 'aws-sdk') // generates 'geocoding-service/aws-sdk'
```
Will it also include the '/**' for hoisting all transitive dependencies?

andrestone replied:
Not by default, will update the README.

eladb commented on README.md:
````diff
@@ -177,11 +177,93 @@ That's it. projen will auto-detect `bin/mycli` and will add it to your
+```typescript
+const { LernaProject, TypeScriptLibraryProject } = require('projen');
+
+const myChildProject =  new TypeScriptLibraryProject({ 
````
Renamed:

```suggestion
const myChildProject =  new TypeScriptProject({ 
```

eladb commented on examples/lerna/.projenrc.ts:
```diff
@@ -0,0 +1,139 @@
+// monorepo
+const cdk8sProject = new LernaProject({
+  name: 'cdk8s',
+	lernaVersioning: false,
```
indentation is off here

eladb commented on package.json:
```diff
@@ -73,7 +71,8 @@
   },
   "license": "Apache-2.0",
-  "version": "0.3.3",
+  "version": "0.3.2",
```
Seems like a downgrade

eladb commented on src/file.ts:
```diff
@@ -63,6 +63,19 @@ export abstract class FileBase extends Construct {
 
+    // Preserving existing field on JSON file (e.g version)
+    // not using JsonFile to avoid circular reference issue
+    if ((this as any).preserveJSONFields?.length) {
```
What's that?

andrestone replied:
If I import `JsonFile` class to be used in the `if` condition (`this instanceof JsonFile`) I get this:
`TypeError: Class extends value undefined is not a constructor or null`

eladb commented on src/jsii-project.ts:
```diff
@@ -137,6 +137,8 @@ export class JsiiProject extends TypeScriptLibraryProject {
 
     this.addFields({ types: `${libdir}/index.d.ts` });
 
+    this.addFields({ main: `${libdir}/index.js` });
```
Let's move these two to the `TypeScriptProject` class.

eladb commented on src/json.ts:
```diff
@@ -2,18 +2,35 @@ import { FileBase, FileBaseOptions } from './file';
+  /**
+   * Fields to preserve from existing file.
+   */
+  readonly preserveJSONFields?: string[];
```
Mmmm. That's an interesting approach. Does it change how we manage versions? Perhaps requires a separate PR?

eladb commented on src/lerna-project.ts:
```diff
@@ -0,0 +1,88 @@
+import * as path from 'path';
+
+export class LernaProject extends NodeProject {
+  public readonly packages: NodeProject[];
```
docstrings for all public APIs please (yes I know I don't have them everywhere, yet, but let's start).

eladb commented on src/node-project.ts:
```diff
@@ -234,6 +240,16 @@ export interface NodeProjectOptions extends ProjectOptions, CommonOptions {
+  /**
+   * Use lerna for versioning
+   */
+  readonly lernaVersioning?: boolean;
```
feels weird that `NodeProject` knows about lerna.

eladb commented on src/node-project.ts:
```diff
@@ -286,21 +302,22 @@ export class NodeProject extends Project {
-        url: options.authorUrl,
-        organization: options.authorOrganization ?? false,
-      },
+      'author': options.authorName ?
```
How is this related to this PR?

eladb commented on src/node-project.ts:
```diff
@@ -312,6 +329,7 @@ export class NodeProject extends Project {
 
     new JsonFile(this, 'package.json', {
       obj: this.manifest,
+      preserveJSONFields: ['version'], // this is managed by another script
```
Extract to a separate PR

eladb commented on src/project.ts:
```diff
@@ -19,7 +19,7 @@ export class Project extends Construct {
-  public synth(): void {
-    Node.of(this).synthesize({ outdir: this.outdir });
+  public synth(outdir?: string): void {
+    Node.of(this).synthesize({ outdir: outdir || this.outdir });
```
```suggestion
    Node.of(this).synthesize({ outdir: outdir ?? this.outdir });
```

eladb commented on src/typescript.ts:
```diff
@@ -94,6 +94,8 @@ export class TypeScriptLibraryProject extends NodeProject {
     this.docgen = options.docgen;
     this.docsDirectory = options.docsDirectory || 'docs/';
 
+    this.addFields({ main: `${this.libdir}/index.js` });
```
the `.d.ts` should go here too and then it's not required in `JsiiProject` anymore

eladb commented on src/version.ts:
```diff
@@ -38,14 +49,29 @@ export class Version extends Construct {
+  }
+}
+
+export class LernaVersion extends Construct {
```
Let's have a separate discussion about versioning and bumping and releases....

eladb commented on templates/jsii/.template/template.js:
```diff
@@ -0,0 +1,9 @@
+const { JsiiProject } = require('projen');
+
+const project = new JsiiProject({
```
please remove the templates

andrestone commented:
Okay.

Couple things that I needed to fix to have it working:

`generateSample()`'s current behaviour is creating files as a side effect of the class instantiation. Refactored it to happen on `project.synth()` so the files are generated according to the `outdir` of the synth session.

Likewise, `Version.current()` was creating an initial `.version.js` file when instanitiated, also refactored so it happens on synth using session `outdir`.

adamelmore commented:
Any plans to finish this out? I'd also like to see Lerna support and I'm happy to help if I can!

andrestone commented:
The project has evolved a lot since this PR was proposed.

It's probably better to start fresh.

adamelmore commented:
@andrestone care if I take this on, or did you still plan to?

andrestone commented:
Please do it!
